design/tab_complete.md

```diff
@@ -0,0 +1,63 @@
+ideally, we could set up a different completion code path that takes advantage of some of the
+additional features provided by zsh.
+* Argument completion. This would mean intelligently completing arguments with DC/OS object
+identifiers, e.g. `dcos cluster attach[TAB]` would give a list of all clusters the CLI is aware of.
+* Flag name completion. Provide completion for flag names available to the current command, e.g.
+`dcos auth list-providers --[TAB]` should give `json`.
+* Flag argument completion. Provide intelligent completion for flag arguments, e.g.
+`dcos node ssh --private-ip=[TAB]` should give a list of the private IPs of all nodes in the cluster.
+
+
+## Implementation ##
+
+To support a mostly static completion script, we're planning to mimic how the
+[Mesos CLI does completion](https://github.com/apache/mesos/blob/master/src/python/cli_new/mesos.bash_completion).
+So for DC/OS it would call something like `dcos __autocomplete__ bash dcos cluster list` and much of the completion
+will be handled by a command built into the CLI. The leading `bash` would indicate which shell-specific completion
+script to execute, it's an argument because users won't be expected to type this regularly since it'll be in the
+completion script which is already shell specific.
+
+For commands that exist outside of the CLI, it will work similarly, shelling out to that command if it supports
+completion, like `<subcommand binary> __autocomplete__ bash marathon task list`. And if they don't support
+`__autocomplete__` then they would need to have some metadata that describes at least some things about the
+binary (subcommands, flags, etc.) so the CLI can provide at least a basic completion experience even on older
+versions of our plugins.
+
+
+## Complications ##
+
+We would like to minimize the amount of special consideration required to make our commands support the
+various types of autocompletion. For this, it would be ideal to use a CLI framework that allows us to look
+at the internals of the commands we've built in order to extract subcommand and flag names. For additional
+pieces like argument completion or flag value completion, we're thinking of adding some metadata to the
+command/flag objects that the autocomplete command can read to determine what special behavior might be needed.
+
+The current problem that we're running into with Cobra comes from how it handles flags. Parsing flags is
+complicated and much of the code to handle it is in private functions within the Cobra package that also
+use fields private to Cobra. This makes it extremely difficult to differentiate a flag followed by a
+subcommand name and a flag that takes a value, e.g. with `dcos -v cluster list` we would need to check
+that `-v` doesn't take a value, therefore `cluster` must be an argument/subcommand. Because of this, it will
+be difficult to do robust argument completion or flag value completion.
```

Comment:
How about not using introspection in order to detect which flags / commands are available in the Go binary? But rather, define this in the code with the structure which makes the most sense for this feature. This would require to keep this structure in sync with the actual command / flags hierarchy, but it's a relatively straightforward task and would avoid the complications you're outlining.

Reply:
This is an interesting thought and now I'm thinking that it makes more sense to abandon cobra and write our own command dispatch.

First, if we're going to maintain our own metadata about the commands and their flags, then really we would end up with a structure that's essentially the same as the completion structure we'll use for external commands so we should just always use the metadata if we go that route.

But this led to me thinking about how we'd implement that and by extension, how we'd deal with external commands. I think in order to parse that structure against the input we've received so far, we'll need to more or less reimplement how cobra does command parsing anyway. The metadata would need to encode what commands there are, the structure of the command tree, what flags each command can take, and whether or not those flags take arguments. We'd also need to be able to parse the given command string to find the appropriate command to get completions for. From there all we need to do is add in a field for the function that the command should run and flag argument info (what type it should be, default values) and we've more or less remade cobra.

Even if we maintained a flattened list of the commands instead of a tree, we'd still need to deal with flags, including flags on parent commands, unless we get rid of those but I think the subcommands might have some.

Reply:
If we don't go with the introspection approach then indeed we'd have to re-implement a subset of what cobra does, but as you said the good part about it is that we would have to do this anyways for plugin subcommands metadata, so there shoud be no wasted work here.

Then whether we move away from cobra to implement our own command dispatcher based on this metadata structure is another topic. I'd need to hear more about that to have a clear opinion, but I feel like a generic declarative structure is not as powerful as writing code and we'd probably face some difficulties if we were to go that route.